### ISSUE TITLE
complete=.,t and CTRL-p sometimes completes keywords from tags before keywords from local file

### ISSUE THREAD

**Issue by knutaf** [vim/vim#1890]
REPRO STEPS:
1. download the attached test.cpp.txt and test_tags.txt
1. edit test.cpp.txt
1. :set cpt=.,t
1. :set tags=tags_test.txt
1. go to the bottom of the file, go into insert mode, and type va followed by pressing CTRL-p repeatedly

EXPECTED: all 100 varNNN keywords are always completed before any valueNNN keywords are completed
ACTUAL: a few varNNN keywords are completed, but then valueNNN keywords start being completed

in contrast, if you type va, then CTRL-p only once and wait for the "scanning tags" message in the status bar to finish, then continue pressing CTRL-p, then the correct priority order of varNNN will be completed before any valueNNN keywords. It seems like when the tag scanning process completes, it inserts tag keyword matches at the current position in the completion list rather than at the end.

You can even see the erroneous behavior if you :set cot=menu. You'll see the incorrect priority order valueNNN keywords get inserted into the middle of the menu list.

[tags_test.txt](https://github.com/vim/vim/files/1183967/tags_test.txt)
[test.cpp.txt](https://github.com/vim/vim/files/1183968/test.cpp.txt)

**Comment by knutaf** [comment]
This worked in vim 7.4.712. Broken in vim 7.4.2367. Not sure where in the middle it broke.

**Comment by chrisbra** [comment]
could you bisect the commit please?

**Comment by knutaf** [comment]
I am not set up to build vim myself, so I just had to look at prebuilt binary versions for Windows. I'm fairly sure the bug was introduced in this change:
https://github.com/vim/vim/commit/67081e50616ae9546621072c5eaaa59bd0a4bed7#diff-14dcf96f167b08fe1c89e75c9444b370

vim 7.4.1964 does not show the bug, and vim 7.4.2015 does show the bug.

**Comment by knutaf** [comment]
Sorry, a bit sleepy here. I forgot I could narrow it further. vim 7.4.2005 does not show the bug. vim 7.4.2015 does show the bug. I still suspect the commit I mentioned in my previous comment.

**Comment by chrisbra** [comment]
hm, I cannot reproduce here. Can you reproduce with `vim -u NONE -N`? What is your 'completeopt' setting (since you suspect 67081e5 to be the fault which is about `:set completeopt+=noinsert`

**Comment by knutaf** [comment]
I can reproduce consistently with vim -u NONE -N on patch 2015 and later, and can consistently not reproduce on patch 2005 and before.

In order to hit the repro, you have to cycle through the first few entries in the completion list before tag scanning finishes. If you hit CTRL-p just once, then wait for tag scanning to finish, then you won't see the bug. For example, if you hit CTRL-p twice rapidly, then hit CTRL-p again, that third time will have an incorrect completion.

My instructions about setting the tags file is crucial. I left completeopt as the default, which is "completeopt=menu,preview". It also reproes consistently with completeopt= (empty).

Any other suggestions on how to help with the repro? Maybe generate a new tags file that has many thousands of valueNNNN entries, to make tag scanning go slower?

**Comment by knutaf** [comment]
Hi, have you had any chance to take a look at this again? I'm confident I can still repro this 100%, and I think I've narrowed it down pretty closely.

**Comment by lifepillar** [comment]
I cannot reproduce this under macOS in any way, even with a much bigger tags file (I have added over a million entries). I have tried 7.4.2005, 7.4.2015, 8.0.946, and they all behave the same: pressing ctrl-p repeatedly goes through var100, var99, …, up to var0, then selects value368, value367, … (Vim shows 401 matches in total—can this limit be lifted?).

I have an SSD, so maybe the issue appears only with a slower disk?

**Comment by lifepillar** [comment]
>I have an SSD, so maybe the issue appears only with a slower disk?

I have just tried with a ~10yr old Linux box with a magnetic disk, and again, I cannot reproduce the issue. Also:

>In order to hit the repro, you have to cycle through the first few entries in the completion list before tag scanning finishes.

Even with a 46MB tags file, the “Scanning tags” message lasts the blink of an eye and it is barely perceptible, if at all. Does it last significantly longer for you?

**Comment by knutaf** [comment]
Hi, thanks for taking a look. I also have an SSD in my laptop, and it's not really a slow laptop (i7, decent specs), but I see "Scanning tags" for about 250 ms pretty consistently. I'd say it's very perceptible. Maybe the slowness is a windows thing. I feel like I see this on any Windows PC, no matter the specs.

I took some comparison videos so you can see what I'm seeing and doing. I did just a plain CTRL-p first to show how long the "Scanning tags" shows on my screen, then showed the buggy completion entries:
vim7.4.2005: https://1drv.ms/u/s!AvoQBZZ9sOA0rpkzO3wL4BkL_TagPw
vim7.4.2015: https://1drv.ms/u/s!AvoQBZZ9sOA0rpk0KNNorH6QRYzGLQ

**Comment by chrisbra** [comment]
could it be, that the tags file is on a network share?

**Comment by knutaf** [comment]
No, everywhere I use tags, it only searches local drive paths. If there's some difference in speed in tag searching between Windows and other OSes, I think it's a separate bug than the incorrect autocompletion entries.

**Comment by chrisbra** [comment]
then perhaps it is a virus scanner intercepting?

**Comment by knutaf** [comment]
I only have the built-in Windows Defender as a scanner. I tried turning it off just now, but there's no difference in tag scanning speed. Regardless of how fast tag scanning goes, for some set of users if tag scanning happens to go too slowly, I think they're going to hit this bug. To repro it, could you try putting your tags file on a network share to slow it down?

**Comment by chrisbra** [comment]
While my main system is actually Windows, I develop only on Linux and don't even have a Windows system available for debugging. But hey, I have the source, so why not patch Vim to be slower :)
```
diff --git a/src/edit.c b/src/edit.c
index 9651c267a..3853f52bd 100644
--- a/src/edit.c
+++ b/src/edit.c
@@ -4382,6 +4382,7 @@ ins_compl_get_exp(pos_T *ini)
                else if (*e_cpt == ']' || *e_cpt == 't')
                {
                    type = CTRL_X_TAGS;
+                   do_sleep(1000);
                    vim_snprintf((char *)IObuff, IOSIZE, _("Scanning tags."));
                    (void)msg_trunc_attr(IObuff, TRUE, HL_ATTR(HLF_R));
                }
```

Now, with that patch, I can reproduce it. However, I cannot promise to fix that. I will look into it a bit, but the completion popup code is kind of hairy and I won't claim to understand every part of it and I don't even have that much spare time currently. So hopefully, this helps other people pinpoint the problem.

**Comment by chrisbra** [comment]
BTW: I can confirm, that 67081e50616ae9546621072c5eaaa59b causes the problem.

**Comment by Shougo** [comment]
Hi, I am the patch author.

> BTW: I can confirm, that 67081e5 causes the problem.

But it fixes https://github.com/vim/vim/issues/874 problem.
So I don't think the simple revert does not fix the problem.

67081e5 causes https://github.com/vim/vim/issues/1168 (it is fixed though).
So I know it is not the smart approach.

**Comment by chrisbra** [comment]
Oh I did not mean to imply, that the offending commit should be reverted. We just should fix the specific problem mentioned here.
